### PR TITLE
wallet: 'CommitTransaction', remove extra wtx lookup and add exception for db write error

### PR DESCRIPTION
**Pull request by furszy**
Two points for `CWallet::CommitTransaction`:

1) The extra wtx lookup:
    As we are calling to `AddToWallet` first, which returns the recently added/updated wtx pointer, there is no need to look up the wtx again few lines later. We can just use it.

2) The db write error:
    `AddToWallet` can only return a nullptr if the db write fails, which inside `CommitTransaction` translates to an exception throw cause. We expect everywhere that `CommitTransaction` always succeed.

------------------------------------------------

Extra note:
This finding generated another working path for me :)
 It starts with the following question: why are we returning a nullptr from `AddToWallet` if the db write failed without removing the recently added transaction from the wallet's map?..
 Can led to a wallet invalid state where the inputs of this new transaction are not marked dirty, while the transaction that spends them still exist on the in-memory wallet tx map.
-- I'm writing it here to gather some feedback first and not forget it, will create a follow-up PR in the coming days 🚜 --

### CODE REVIEW

**Comment by achow101** [comment]
ACK 57fb37c27599fc865f20b42a27bb9c227f384de3

**Review by jonatack — COMMENTED** [review]
ACK 57fb37c 

One question regarding the exception.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2116,24 +2116,25 @@ void CWallet::CommitTransaction(CTransactionRef tx, mapValue_t mapValue, std::ve
         return true;
     });
 
+    // wtx can only be null if the db write failed.
+    if (!wtx) {
+        throw std::runtime_error(std::string(__func__) + ": Wallet db error, transaction commit failed");
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/wallet.cpp:2121]: > Might be simpler and easier to maintain if this is raised directly at the source?

At least for now (with the current code architecture), as `AddToWallet` is called from `CommitTransaction`, `importPrunedFunds` and  `AddToWalletIfInvolvingMe` (and this last one is used in flows like chain sync and chain rescan), would keep it as is so we are able to throw different error descriptions on the method's caller side.
User and debugging wise, isn't the same to throw an error for a failure in the transaction commit flow than a failure storing a transaction when the node is syncing or rescanning the chain. They should be presented differently to the user and logs.

Nevertheless, I agree that it's a good point to take into account moving forward 👍🏼. We could improve the error handling returning a db failure status instead of the obscure `nullptr`.. (which originated other problematic -> #25272).

> Edit: Would it be helpful to indicate with which tx the db write failed?

I don't think that is needed on this specific process. `CommitTransaction` is trigger directly from a user action and the user does not know which is the id of the tx that the wallet is creating for him/her (he/she just pressed a button in the GUI or called to RPC commands like `sendtoaddress`, `send`, `sendall`, etc).
In other words, as a user, if I press the send button on the GUI (or call `sendtoaddress`), and the process fails, the tx id on the error message isn't helpful for me. It would actually be confusing.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 57fb37c27599fc865f20b42a27bb9c227f384de3. Seems like a clear improvement. Better to fail earlier with a better error message if the failure is going to happen anyway
